R/methods-CollectionWithMetadata.R

```diff
@@ -0,0 +1,247 @@
+setMethod("removeIncompleteRecords", signature("CollectionWithMetadata"), function(object, colName = character(), verbose = c(TRUE, FALSE)) {
+    verbose <- veupathUtils::matchArg(verbose)
+
+    df <- getCollectionValues(object, verbose = verbose)
+    sampleMetadata <- getSampleMetadata(object)
+
+    # Remove Records with NA from data and metadata
+    if (any(is.na(sampleMetadata[[colName]]))) {
+        veupathUtils::logWithTime("Found NAs in specified variable. Removing these records.", verbose)
+        recordsWithData <- which(!is.na(sampleMetadata[[colName]]))
+        # Keep records with data. Recall the CollectionWithMetadata object requires records to be in the same order
+        # in both the data and metadata
+        sampleMetadata <- sampleMetadata[recordsWithData, ]
+        df <- df[recordsWithData, ]
+
+        object@data <- df
+        object@sampleMetadata <- SampleMetadata(
+            data = sampleMetadata,
+            recordIdColumn = object@sampleMetadata@recordIdColumn
+        )
+        validObject(object)
+    }
+
+    return(object)
+})
+
+#' Prune features by predicate
+#' 
+#' Modifies the data slot of an 
+#' CollectionWithMetadata object, to exclude features for which 
+#' the provided predicate function returns FALSE.
+#' 
+#' @param object CollectionWithMetadata
+#' @param predicate Function returning a boolean indicating if a feature should be included (TRUE) or excluded (FALSE)
+#' @param verbose boolean indicating if timed logging is desired
+#' @return CollectionWithMetadata with modified data slot
+#' @rdname pruneFeatures
+#' @export
+setGeneric("pruneFeatures",
+    function(object, predicate, verbose = c(TRUE, FALSE)) standardGeneric("pruneFeatures"),
+    signature = c("object")
+)
+
+#' @rdname pruneFeatures
+#' @aliases pruneFeatures,CollectionWithMetadata-method 
+setMethod("pruneFeatures", signature("CollectionWithMetadata"), function(object, predicate, verbose = c(TRUE, FALSE)) {
+    df <- getCollectionValues(object)
+    allIdColumns <- c(object@recordIdColumn, object@ancestorIdColumns)
+
+    # keep columns that pass the predicate
+    keepCols <- df[, lapply(.SD, predicate), .SDcols = colnames(df)[!(colnames(df) %in% allIdColumns)]]
+    keepCols <- names(keepCols)[keepCols == TRUE]
+    df <- df[, c(allIdColumns, keepCols), with = FALSE]
+
+    # LET ME EXPLAIN
+    # since we called getCollectionValues (which removes empty records)..
+    # we need to do the same for sampleMetadata in order to produce a valid object.
+    # and, we dont want those empty records influencing which features get pruned, so i think were tied to this.
+    # we just need to be sure we ask for the metadata before resetting the abundance data, or else we'll get an error
+    # bc getSampleMetadata also calls getCollectionValues to find which records to remove
+    # we could maybe do it better, by introducing a hasSampleMetadata method in here. but i'm not sure if that's worth it.
```

Comment:
why not have a Collection method called removeEmptyRecorsds(collection) and then that's what's called first? 
Like 
```
object <- removeEmptyRecords(object)
df <- getCollectionValues(object)
```

Then the sample metadata will already be trimmed. And the pruning of the df should have no influence on the metadata, because we're removing columns not rows, no?